### PR TITLE
bpo-41873: Add vectorcall for float()

### PR DESCRIPTION
**Pull request by sweeneyde**
https://bugs.python.org/issue41873

### CODE REVIEW

**Review by corona10 — CHANGES_REQUESTED** [review]
Your patch will cause the crash
```
Type "help", "copyright", "credits" or "license" for more information.
>>> float()
[1]    44448 segmentation fault  ./python.exe
```

**Review by corona10 — COMMENTED** [review]
~Please add the test case that I pointed out~ -> the case is already existed.

File: Objects/floatobject.c [hunk 1]
```diff
@@ -1649,6 +1649,23 @@ float_subtype_new(PyTypeObject *type, PyObject *x)
...
+    if (!_PyArg_NoKwnames("float", kwnames)) {
+        return NULL;
+    }
+
+    Py_ssize_t nargs = PyVectorcall_NARGS(nargsf);
+    if (!_PyArg_CheckPositional("float", nargs, 0, 1)) {
+        return NULL;
+    }
+
+    return float_new_impl((PyTypeObject *)type, args[0]);
```
Comment (corona10) [inline Objects/floatobject.c]:
```suggestion
    PyObject *x = NULL;
    if (nargs > 0) {
        x = args[0];
    } else {
        x = _PyLong_Zero;
    }
    return float_new_impl((PyTypeObject *)type, x);
```

**Comment by sweeneyde** [comment]
I have made the requested changes; please review again.

@corona10 I didn't see any occurrences of ``float()`` in test_float.py, so I added a test. Where are you saying the test already exists?

**Comment by corona10** [comment]
> I didn't see any occurrences of float() in test_float.py

https://github.com/python/cpython/pull/22432/checks?check_run_id=1175574849

There was already a crash on your PR on CI phase.
So I guess that there is already in, but it looks like good to add the case on `test_float.py`

**Review by corona10 — APPROVED** [review]
I am lgtm about the changes,
but let's wait for other core devs opinions about adding vectorcall for float type.

**Comment by corona10** [comment]
@markshannon 

This PR is same case from https://github.com/python/cpython/pull/22427

Do you think that the vectorcall is good enough to apply for float() call also from the maintainence cost view?

**Review by vstinner — APPROVED** [review]
LGTM. IMHO removing 27.5 ns on 79 ns (1.5x faster) per float(obj) is worth it.

I let @corona10 merge the PR ;-)

**Pull request merged**